src/main.cpp

```diff
@@ -2871,6 +2881,22 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
         }
     }
 
+    // History read/write is unconditional, since it is generated for entire
+    // chain regardless of the Heartwood upgrade activation height.
+    {
+        auto historyNode = libzcash::NewLeaf(
+            block.GetHash(), 
+            block.nTime, 
+            block.nBits, 
+            block.hashFinalSaplingRoot,
+            ArithToUint256(pindex->nChainWork),
+            pindex->nHeight,
+            total_shielded_tx
+        );
+
+        view.PushHistoryNode(historyNode);
```

Comment:
Still todo: handle crossing a network upgrade activation boundary (where the tree is restarted).